### ISSUE TITLE
Embedding within Parallel:Loops

### ISSUE THREAD

**Issue by palakpsheth** [lskatz/Schedule--SGELK#1]
Hi Lee,

Thank you for very much for a very useful module! I am having an issue when trying to embed within a Parallel::Loops foreach loop. I want to queue up at most N jobs at any one time. When I set the parallel loops to say 4, the expected behavior is to process the first 4 items in the array submit then them as jobs then wait. Once one job is finished, the next in the array starts. This way I would never have more than 4 jobs running/queued at any one time. 

I am finding that when I set this up, each of the 4 parallel loops are calling the same qsub script that gets generated. I have confirmed that indeed the first 4 items of the array are getting parsed but only the last item has a script generated and gets submitted.

Any ideas would be a great help. Thanks!

Attached is my main wrapper script that contains the parallel loops along with the SGE scheduling.

Thanks again for any help you can provide.

Regards,
Palak
[generate_RAWCOUNTS_SGE - Copy.txt](https://github.com/lskatz/Schedule--SGELK/files/360466/generate_RAWCOUNTS_SGE.-.Copy.txt)

**Comment by lskatz** [comment]
I think that your script could be overwriting the files under `$tempdir`.  You should set a new temporary directory each time I think.  For example, and this is untested:

```
use Schedule::SGELK;
use File::Tempdir qw/tempdir/;
use File::Basename qw/basename/;
use File::Spec::Functions qw/catdir/;

$tempDir = catdir($outDir,"temp");

$pl->foreach (\@fastqFilePaths, sub {
  $subTempdir=tempdir("$tempdir/".basename($0)."XXXXXX",TMPDIR=>0,CLEANUP=>0);
  $command=".... --tempdir $subTempdir ...";
  $sge->pleaseExecute($command);
}
```

Please let me know if this fixes your script, so that I can close it.  Thank you for using `SGELK`!

**Comment by palakpsheth** [comment]
Thanks!

You are correct that the script files were being overwritten. This is because the "random" number that was being used to name the qsub script was the same for the parallel threads. This is because perl rand() function is based on a seed. The default seed is the datetime so all the parallel jobs launched at same datetime got the same random int. 

I resolved the issue by modifying the qsub script naming to include the jobname as part of the filename string. This work for me since I am uniquely naming the jobs that are spawned in parallel. I also enabled passing in an additional optional seed value.

I am attaching here for your consideration. If you would like I can fork and send a pull request.
[SGELK.pm.txt](https://github.com/lskatz/Schedule--SGELK/files/362873/SGELK.pm.txt)
